src/cluster/migrate_iterator.cc

```diff
@@ -0,0 +1,159 @@
+ * under the License.
+ *
+ */
+
+#include "cluster/migrate_iterator.h"
+
+#include "db_util.h"
+#include "storage/redis_db.h"
+
+MigrateIterator::MigrateIterator(engine::Storage *storage, const rocksdb::ReadOptions &read_options)
+    : metadata_cf_(storage->GetCFHandle(engine::kMetadataColumnFamilyName)),
+      subkey_cf_(storage->GetCFHandle(engine::kSubkeyColumnFamilyName)),
+      zset_score_cf_(storage->GetCFHandle(engine::kZSetScoreColumnFamilyName)),
+      stream_cf_(storage->GetCFHandle(engine::kStreamColumnFamilyName)),
+      metadata_iter_(util::UniqueIterator(storage, read_options, metadata_cf_)),
+      subdata_iter_(util::UniqueIterator(storage, read_options, subkey_cf_)),
+      stream_iter_(util::UniqueIterator(storage, read_options, stream_cf_)),
+      valid_(false),
+      metadata_(RedisType::kRedisNone, false) {}
+
+bool MigrateIterator::Valid() const { return valid_; }
+
+void MigrateIterator::Seek(const rocksdb::Slice &target) {
+  items_.clear();
+  log_data_.clear();
+  metadata_iter_->Reset();
+  subdata_iter_->Reset();
+  stream_iter_->Reset();
+  metakey_prefix_.clear();
+
+  metadata_iter_->Seek(target);
+  valid_ = metadata_iter_->Valid() && metadata_iter_->key().starts_with(target);
+  if (valid_) {
+    metakey_prefix_ = target.ToString();
+    findMetaData();
+  }
+}
+
+void MigrateIterator::Next() {
+  assert(valid_);
+  valid_ = false;
+  items_.clear();
+
+  if (subdata_iter_->Valid()) {
+    subdata_iter_->Next();
+    valid_ = subdata_iter_->Valid() && subdata_iter_->key().starts_with(subkey_prefix_);
+    if (!valid_) {
+      subdata_iter_->Reset();
+    }
+  } else if (stream_iter_->Valid()) {
+    stream_iter_->Next();
+    valid_ = stream_iter_->Valid() && stream_iter_->key().starts_with(subkey_prefix_);
+    if (!valid_) {
+      stream_iter_->Reset();
+    }
+  }
+
+  if (valid_) {
+    findSubData();
+  } else if (metadata_iter_->Valid()) {
+    metadata_iter_->Next();
+    valid_ = metadata_iter_->Valid() && metadata_iter_->key().starts_with(metakey_prefix_);
+    if (valid_) {
+      findMetaData();
+    }
+  }
+}
+
+const std::vector<MigrateItem> &MigrateIterator::GetItems() const {
+  assert(valid_);
+  return items_;
+}
+
+std::string MigrateIterator::GetLogData() const {
+  assert(valid_);
+  return log_data_;
+}
+
+void MigrateIterator::findMetaData() {
+  assert(metadata_iter_->Valid());
+  Metadata metadata(kRedisNone, false /* generate_version */);
+  metadata.Decode(metadata_iter_->value().ToString());
+  RedisType redis_type = metadata.Type();
+  metadata_ = metadata;
+
+  redis::WriteBatchLogData log_data(redis_type);
+  if (redis_type == RedisType::kRedisList) {
+    log_data.SetArguments({std::to_string(RedisCommand::kRedisCmdRPush)});
+  }
+
+  log_data_ = log_data.Encode();
+
+  items_.push_back(MigrateItem{metadata_cf_, metadata_iter_->key().ToString(), metadata_iter_->value().ToString()});
+
+  if (redis_type == RedisType::kRedisStream) {
```

Comment:
If we can move this reset inside the initSubData?

Reply:
Yes, it looks fine